art/attacks/evasion/auto_projected_gradient_descent.py

```diff
@@ -275,7 +215,24 @@ def __call__(self, y_true, y_pred):
                             "the estimator has to to predict logits."
                         )
 
-                    self._loss_object = torch.nn.CrossEntropyLoss(reduction="mean")
+                    # modification for image-wise stepsize update
+                    class CrossEntropyLossTorch:
+                        """Class defining cross entropy loss with reduction options."""
+
+                        def __init__(self, reduction="mean"):
+                            self.ce_loss = torch.nn.CrossEntropyLoss(reduction="none")
+                            self.reduction = reduction
+
+                        def __call__(self, y_true, y_pred):
+                            if self.reduction == "mean":
+                                return self.ce_loss(y_true, y_pred).mean()
+                            if self.reduction == "sum":
+                                return self.ce_loss(y_true, y_pred).sum()
+                            if self.reduction == "none":
+                                return self.ce_loss(y_true, y_pred)
+                            raise NotImplementedError()
+
+                    self._loss_object = CrossEntropyLossTorch(reduction="mean")
                 elif loss_type == "difference_logits_ratio":
                     if is_probability(
                         estimator.predict(x=np.ones(shape=(1, *estimator.input_shape), dtype=ART_NUMPY_DTYPE))
```

Comment:
Please change line 245 from `class DifferenceLogitsRatioPyTorch:` to `class DifferenceLogitsRatioPyTorch(torch.nn.modules.loss._Loss):`